app/services/github/oauth_client.rb

```diff
@@ -0,0 +1,135 @@
+module Github
+  # Github OAuth2 client (uses ocktokit.rb as a backend)
+  class OauthClient
+    APP_AUTH_CREDENTIALS = %i[client_id client_secret].freeze
+
+    # @param credentials [Hash] the OAuth credentials, {client_id:, client_secret:} or {access_token:}
+    def initialize(credentials)
+      @credentials = check_credentials!(credentials)
+    end
+
+    def self.for_user(user)
+      access_token = user.identities.github.select(:token).take!.token
+      new(access_token: access_token)
+    end
+
+    # Hides private credentials when printed
+    def inspect
+      "#<#{self.class.name}:#{object_id}>"
+    end
+
+    private
+
+    attr_reader :credentials
+
+    # adapted from https://api.rubyonrails.org/classes/Module.html#method-i-delegate_missing_to
+    def method_missing(method, *args, &block)
+      return super unless target.respond_to?(method, false)
+
+      # define for re-use
+      self.class.define_method(method) do |*new_args, &new_block|
+        request do
+          target.public_send(method, *new_args, &new_block)
+        end
+      end
+
+      # call the original method, this will only be called the first time
+      # as in subsequent calls, the newly defined method will prevail
+      request do
+        target.public_send(method, *args, &block)
+      end
+    end
+
+    # adapted from https://api.rubyonrails.org/classes/Module.html#method-i-delegate_missing_to
+    def respond_to_missing?(method, _include_all = false)
+      target.respond_to?(method, false) || super
+    end
+
+    def check_credentials!(credentials)
+      if credentials.present?
+        return credentials.select { |k| k == :access_token } if credentials[:access_token].present?
+
+        all_auth_credentials = proc { |key, value| APP_AUTH_CREDENTIALS.include?(key) && value.present? }
+        return credentials.select { |key| APP_AUTH_CREDENTIALS.include?(key) } if credentials.all?(all_auth_credentials)
```

Comment:
Am I misreading this, or could you  

```ruby
return credentials if credentials.all?(all_auth_credentials)
```

since the condition already verifies that all the keys are in `APP_AUTH_CREDENTIALS`?

I think actually all of this could be a bit simplified (untested):

```ruby
# assumes we move the proc into a constant
result = credentials[:access_token] || (credentials if credentials.all?(ALL_AUTH_CREDENTIALS))
return result if result
```

Reply:
Good idea! will update it shortly